### DISCUSSION TITLE
Updating GitHub actions runners

### DISCUSSION THREAD

**Discussion by DanySK** [renovatebot/renovate#15213] [Suggest an Idea]
Hi,
is there a way to configure renovate to propose updates to GitHub actions runners? Many use "xxx-latest" environments, but that makes the build hard to reproduce in the future.
I'd like to fix the version, but I don't want to fall behind when new runners are released - it's a dependency like any other.

Also, it would be nice if it worked in build matrices.

**Comment by viceice** [comment]
the syntax is too complex to handle them buildin, so i suggest to use the regex manager for it.

one problem it to find an appropriate datasource to fetch updates from.

**Reply by DanySK** [reply]
@viceice Well, I think that I can match quite easily using this... https://docs.renovatebot.com/modules/manager/regex/
Indeed, the issue is with the datasource.
I can easily scrape the `README.md` here: https://github.com/actions/virtual-environments and extract updates, but I need a datasource capable of doing this.
The two options that come to my mind are:
1. A general "webpage-scraper" datasource, that fetches a URL, matches a regex, and extracts updates
2. A new "github-hosted-runner' datasource, specifically supporting the sole case of GHA

The former would be preferable from an engineering POV (also, I guess some other datasources could be just instances of that), but I can see it exploited too much. The latter would be quicker, easier to maintain, and not exploitable.
Would any of the two be accepted as contribution? I do not want to start writing pieces of software if there are no guarantees that the idea is okay with the maintainers.

**Reply by viceice** [reply]
there is an issue for a generic datasource

**Reply by rarkins** [reply]
Possibilities:
- Create a new repo with a cron'd Action which scrapes the data and converts it to Releases. Then use that as your datasource
- #10622 is along similar lines but I don't think is applicable here
- #6223 is another idea (similar to your 1 above?)

**Reply by TWiStErRob** [reply]
* This was implemented in https://github.com/renovatebot/renovate/issues/19827

**Comment by DanySK** [comment]
I think I will go for the suggested workaround: generate a repo that hosts the information as fetched from that page into Releases. I'll need three repos, though, if I want to support Linux/Mac/Win -- am I correct?

Yeah, #6223 is definitely similar to what I proposed. I think that 10622 could help as well, if it can fetch any URL, it would boil the problem down to maintaining a publicly available JSON/YAML file with the information in that page.

**Reply by rarkins** [reply]
> I think I will go for the suggested workaround: generate a repo that hosts the information as fetched from that page into Releases. I'll need three repos, though, if I want to support Linux/Mac/Win -- am I correct?

Depends. You could possibly use a release suffix to merge them into one repo and then `extractVersion` to separate them.

If you go this path, would be great if you can somehow document/publish/describe it for others to learn from in future!

**Reply by DanySK** [reply]
I need to study how that datasource works -- is there any specific documentation?

**Reply by rarkins** [reply]
`github-releases` is the datasource I had in mind. `extractVersion` was the feature I suggested for how to filter releases using suffix

**Reply by DanySK** [reply]
I think that I got it.

```json
  "packageRules": [
    {
      "matchDepNames": "windows github actions runner",
      "extractVersion": "^win(?<version>\\d\\d)\/\\d+.\\.\\d+$"
    }
  ],
  "regexManagers": [
    {
      "fileMatch": [
        "^(workflow-templates|\\.github\\/workflows)\\/[^/]+\\.[Yy][Aa]?[Mm][Ll]$",
        "(^|\\/)action\\.[Yy][Aa]?[Mm][Ll]$"
      ],
      "matchStrings": [
          "windows-\\d\\d(?<currentValue>\\d\\d)"
      ],
      "depNameTemplate": "windows github actions runner",
      "packageNameTemplate": "actions/runner-images",
      "datasourceTemplate": "github-releases",
      "versioningTemplate": "docker"
    }
  ]
```

https://github.com/DanySK/Template-for-Kotlin-JVM-Projects/pull/335

**Reply by DanySK** [reply]
Got it working for all three runners

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "github>DanySK/renovate-config:kotlin"
  ],
  "packageRules": [
    {
      "matchDepNames": "windows github actions runner",
      "extractVersion": "^win(?<version>\\d\\d)\/\\d+.\\.\\d+$"
    },
    {
      "matchDepNames": "macos github actions runner",
      "extractVersion": "^macOS-(?<version>\\d\\d)\/\\d+.\\.\\d+$"
    },
    {
      "matchDepNames": "ubuntu github actions runner",
      "extractVersion": "^ubuntu(?<version>\\d\\d)\/\\d+.\\.\\d+$"
    }
  ],
  "regexManagers": [
    {
      "fileMatch": [
        "^(workflow-templates|\\.github\\/workflows)\\/[^/]+\\.[Yy][Aa]?[Mm][Ll]$",
        "(^|\\/)action\\.[Yy][Aa]?[Mm][Ll]$"
      ],
      "matchStrings": [
          "windows-\\d\\d(?<currentValue>\\d\\d)"
      ],
      "depNameTemplate": "windows github actions runner",
      "packageNameTemplate": "actions/runner-images",
      "datasourceTemplate": "github-releases",
      "versioningTemplate": "docker"
    },
    {
      "fileMatch": [
        "^(workflow-templates|\\.github\\/workflows)\\/[^/]+\\.[Yy][Aa]?[Mm][Ll]$",
        "(^|\\/)action\\.[Yy][Aa]?[Mm][Ll]$"
      ],
      "matchStrings": [
          "macos-(?<currentValue>\\d\\d)"
      ],
      "depNameTemplate": "macos github actions runner",
      "packageNameTemplate": "actions/runner-images",
      "datasourceTemplate": "github-releases",
      "versioningTemplate": "docker"
    },
    {
      "fileMatch": [
        "^(workflow-templates|\\.github\\/workflows)\\/[^/]+\\.[Yy][Aa]?[Mm][Ll]$",
        "(^|\\/)action\\.[Yy][Aa]?[Mm][Ll]$"
      ],
      "matchStrings": [
          "ubuntu-(?<currentValue>\\d\\d)\\.04"
      ],
      "depNameTemplate": "ubuntu github actions runner",
      "packageNameTemplate": "actions/runner-images",
      "datasourceTemplate": "github-releases",
      "versioningTemplate": "docker"
    }
  ]
}
```

**Reply by TWiStErRob** [reply]
Note for anyone finding this in the future:

GitHub Actions manager supports GitHub Actions Runners upgrades automatically:
https://docs.renovatebot.com/modules/manager/github-actions/
